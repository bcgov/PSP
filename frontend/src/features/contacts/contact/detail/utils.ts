import { IContactAddress } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
