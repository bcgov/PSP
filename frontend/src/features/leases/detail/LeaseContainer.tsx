import { SidebarStateContext } from 'components/layout/SideNavBar/SideNavbarContext';
import { SidebarContextType } from 'components/layout/SideNavBar/SideTray';
import LoadingBackdrop from 'components/maps/leaflet/LoadingBackdrop/LoadingBackdrop';
import queryString from 'query-string';
import * as React from 'react';
import { ReactElement, useContext } from 'react';
import { useLocation } from 'react-router-dom';

import {
  BackToSearchButton,
  LeaseBreadCrumb,
  LeaseHeader,
  LeaseIndex,
  LeaseLayout,
  LeasePageForm,
  useLeaseDetail,
} from '..';
import Details from './LeasePages/details/Details';
import Tenant from './LeasePages/tenant/Tenant';

export interface ILeaseAndLicenseContainerProps {
  match?: any;
}

/**
 * Top level container for lease details, provides logic for loading and controlling lease display
 * @param {ILeaseAndLicenseContainerProps} props
 */
export const LeaseContainer: React.FunctionComponent<ILeaseAndLicenseContainerProps> = props => {
  const { setTrayPage } = useContext(SidebarStateContext);
  const onClickManagement = () => setTrayPage(SidebarContextType.LEASE);

  const { search } = useLocation();
  const { leasePageName } = queryString.parse(search);
  const { lease } = useLeaseDetail(props?.match?.params?.leaseId);

  const leasePage = leasePages.get(leasePageName as LeasePageNames);
  if (!leasePage) {
    throw Error('The requested lease page does not exist');
  }
  return (
    <>
      <LeaseLayout>
        <LeaseBreadCrumb
          leaseId={props.match.leaseId}
          leasePage={leasePage}
          onClickManagement={onClickManagement}
        />
        <LeaseHeader lease={lease} />
        <BackToSearchButton />
        <LeaseIndex currentPageName={leasePageName} leaseId={lease?.id}></LeaseIndex>
        <LeasePageForm leasePage={leasePage} lease={lease}></LeasePageForm>
      </LeaseLayout>
      <LoadingBackdrop show={!lease} />
    </>
  );
};

export interface ILeasePage {
  component: ReactElement;
  title: string;
  description?: string;
}

export enum LeasePageNames {
  DETAILS = 'details',
  TENANT = 'tenant',
  PAYMENTS = 'payments',
  IMPROVEMENTS = 'improvements',
  INSURANCE = 'insurance',
  DEPOSIT = 'deposit',
  SECURITY = 'security',
  SURPLUS = 'surplus',
}

export const leasePages: Map<LeasePageNames, ILeasePage> = new Map<LeasePageNames, ILeasePage>([
  [LeasePageNames.DETAILS, { component: <Details />, title: 'Details' }],
  [
    LeasePageNames.TENANT,
    {
      component: <Tenant />,
      title: 'Tenant',
      description: 'The following is information related to the leasee or licensee',
    },
  ],
  [LeasePageNames.PAYMENTS, { component: <></>, title: 'Payments' }],
  [LeasePageNames.IMPROVEMENTS, { component: <></>, title: 'Improvements' }],
  [LeasePageNames.INSURANCE, { component: <></>, title: 'Insurance' }],
  [LeasePageNames.DEPOSIT, { component: <></>, title: 'Deposit' }],
  [LeasePageNames.SECURITY, { component: <></>, title: 'Physical Security' }],
  [LeasePageNames.SURPLUS, { component: <></>, title: 'Surplus Declaration' }],
]);

export default LeaseContainer;
