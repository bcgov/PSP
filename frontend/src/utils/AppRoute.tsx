import React from 'react';
import { Route, RouteComponentProps, RouteProps } from 'react-router-dom';
import PrivateRoute from 'utils/PrivateRoute';

interface BaseAppRoute extends RouteProps {
  protected?: boolean;
  role?: string | string[];
  claim?: string | string[];
  title: string;
}
interface ComponentRoute extends BaseAppRoute {
  customComponent: React.ComponentType<any>;
  layout?: React.ComponentType<any>;
}

interface RenderRoute extends BaseAppRoute {
  customRender: (props: RouteComponentProps<any>) => React.ReactNode;
}
type IAppRouteProps = ComponentRoute | RenderRoute;

function isRenderRoute(route: RenderRoute | ComponentRoute): route is RenderRoute {
  return (route as RenderRoute).customRender !== undefined;
}

const AppRoute: React.FC<IAppRouteProps> = props => {
  document.title = props.title;

  if (!!props.protected) {
    if (isRenderRoute(props)) {
      return (
        <PrivateRoute
          {...props}
          customRender={props.customRender}
          role={props.role}
          claim={props.claim}
        />
      );
    } else {
      const Layout =
        props.layout === undefined ? (props: any) => <>{props.children}</> : props.layout;
      return (
        <PrivateRoute
          {...props}
          customComponent={props.customComponent}
          layout={Layout}
          role={props.role}
          claim={props.claim}
        />
      );
    }
  }

  if (isRenderRoute(props)) {
    return <Route {...props} render={props.customRender} />;
  } else {
    const Component = props.customComponent;
    const Layout =
      props.layout === undefined ? (props: any) => <>{props.children}</> : props.layout;
    return (
      <Route
        {...props}
        render={props => (
          <Layout>
            <Component {...props} />
          </Layout>
        )}
      />
    );
  }
};

export default AppRoute;
