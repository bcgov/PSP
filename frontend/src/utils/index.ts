export * from './getCurrentFiscal';
export * from './getCurrentYearEvaluation';
export * from './getMostRecentAppraisal';
export * from './getMostRecentEvaluation';
export * from './mapLookupCode';
export * from './mockKeycloak';
export * from './numberFormatUtils';
export * from './propertyUtils';
export * from './utils';
