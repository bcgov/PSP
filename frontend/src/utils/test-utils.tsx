import { useKeycloak } from '@react-keycloak/web';
import {
  fireEvent,
  render as rtlRender,
  RenderOptions as RtlRenderOptions,
  RenderResult,
} from '@testing-library/react';
import { FilterProvider } from 'components/maps/providers/FIlterProvider';
import { createMemoryHistory, MemoryHistory } from 'history';
import noop from 'lodash/noop';
import React, { ReactNode } from 'react';
import { MapContainer } from 'react-leaflet';
import { Router } from 'react-router-dom';
import { ToastContainer } from 'react-toastify';

import TestCommonWrapper from './TestCommonWrapper';

export const mockKeycloak = (
  props: {
    claims?: string[];
    roles?: string[];
    organizations?: number[];
    authenticated?: boolean;
  } = {},
) => {
  const { claims, roles, organizations, authenticated = true } = props;
  // mock what would be returned by keycloak userinfo endpoint
  const userInfo = {
    organizations: organizations ?? [1],
    roles: claims ?? [],
    groups: roles ?? [],
    email: 'test@test.com',
    name: 'Chester Tester',
  };

  (useKeycloak as jest.Mock).mockReturnValue({
    keycloak: {
      userInfo,
      subject: 'test',
      authenticated,
      loadUserInfo: jest.fn().mockResolvedValue(userInfo),
    },
  });
};

export const fillInput = async (
  container: HTMLElement,
  name: string,
  value: any,
  type: string = 'input',
) => {
  let input: Element | null = null;

  if (type === 'radio') {
    input = container.querySelector(`#input-${name}`);
  } else {
    if (type === 'typeahead' || type === 'datepicker') {
      input = container.querySelector(`input[name="${name}"]`);
    } else {
      input = container.querySelector(`${type}[name="${name}"]`);
    }
  }

  // abort early if no input field found
  if (!input) return { input };
  if (type === 'input') {
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    fireEvent.focusOut(input);
  } else if (type === 'typeahead') {
    fireEvent.focus(input);
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    const select = container.querySelector(`[aria-label="${value}"]`);
    fireEvent.click(select!);
    fireEvent.focusOut(input);
  } else if (type === 'datepicker') {
    fireEvent.mouseDown(input);
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    fireEvent.keyPress(input, { key: 'Enter', code: 'Enter' });
  } else if (type === 'radio') {
    fireEvent.click(input);
    fireEvent.focusOut(input);
  } else {
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    fireEvent.focusOut(input);
  }
  fireEvent.blur(input);

  return { input };
};

export const flushPromises = () => new Promise(setImmediate);

export const deferred = () => {
  let resolve: (value?: unknown) => void = noop;
  const promise = new Promise(_resolve => {
    resolve = _resolve;
  });
  return {
    resolve,
    promise,
  };
};

/**
 * Utility type for generic props of component testing
 */
export interface PropsWithChildren {
  children?: ReactNode;
}

const defaultHistory = createMemoryHistory({
  getUserConfirmation: (message, callback) => callback(true),
});

/**
 * Creates a in-memory router for unit testing
 * @param history (optional) a memory history to use instead of default
 * @returns The route provider
 */
export function createRouteProvider(history?: MemoryHistory) {
  return function Wrapper({ children }: PropsWithChildren) {
    return <Router history={history ?? defaultHistory}>{children}</Router>;
  };
}

/**
 * Creates a Map wrapper for unit testing
 * @param done A callback that will be called when the map has finished rendering
 * @returns The map container instance
 *
 * @example
 *    const { promise, resolve } = deferred();
 *    render(<TestComponent/>, { wrapper: createMapContainer(resolve) })
 *    await waitFor(() => promise)
 *    // the map is fully initialized here...
 */
export function createMapContainer(
  done: () => void = noop,
  whenCreated: (map: L.Map) => void = noop,
) {
  return function Container({ children }: PropsWithChildren) {
    return (
      <div id="mapid" style={{ width: 500, height: 500 }}>
        <MapContainer
          center={[48.43, -123.37]}
          zoom={14}
          whenReady={done}
          whenCreated={whenCreated}
        >
          {children}
        </MapContainer>
      </div>
    );
  };
}

// re-export everything from RTL
export * from '@testing-library/react';

// override RenderOptions interface
export interface RenderOptions extends RtlRenderOptions {
  store?: any;
  history?: MemoryHistory;
  useMockAuthentication?: boolean;
  organizations?: number[];
  claims?: string[];
  roles?: string[];
}

function render(
  ui: React.ReactElement,
  options: Omit<RenderOptions, 'wrapper'> = {},
): RenderResult {
  const {
    store,
    history,
    useMockAuthentication = false,
    organizations,
    claims,
    roles,
    ...renderOptions
  } = options;

  // mock authentication state prior to rendering. Check first that keycloak has been mocked!
  if (!!useMockAuthentication || !!claims || !!roles || !!organizations) {
    if (typeof (useKeycloak as jest.Mock).mockReturnValue === 'function') {
      mockKeycloak({
        claims: claims ?? [],
        roles: roles ?? [],
        organizations: organizations ?? [1],
        authenticated: true,
      });
    }
  }

  function AllTheProviders({ children }: PropsWithChildren) {
    return (
      <TestCommonWrapper store={store} history={history}>
        <ToastContainer
          autoClose={5000}
          hideProgressBar
          newestOnTop={false}
          closeOnClick={false}
          rtl={false}
          pauseOnFocusLoss={false}
        />
        <FilterProvider>{children}</FilterProvider>
      </TestCommonWrapper>
    );
  }
  return rtlRender(ui, { wrapper: AllTheProviders, ...renderOptions });
}

async function renderAsync(
  ui: React.ReactElement,
  options: Omit<RenderOptions, 'wrapper'> = {},
): Promise<RenderResult> {
  const {
    store,
    history,
    useMockAuthentication = false,
    organizations,
    claims,
    roles,
    ...renderOptions
  } = options;

  // mock authentication state prior to rendering. Check first that keycloak has been mocked!
  if (!!useMockAuthentication || !!claims || !!roles || !!organizations) {
    if (typeof (useKeycloak as jest.Mock).mockReturnValue === 'function') {
      mockKeycloak({
        claims: claims ?? [],
        roles: roles ?? [],
        organizations: organizations ?? [1],
        authenticated: true,
      });
    }
  }

  function AllTheProviders({ children }: PropsWithChildren) {
    return (
      <TestCommonWrapper store={store} history={history}>
        <ToastContainer
          autoClose={5000}
          hideProgressBar
          newestOnTop={false}
          closeOnClick={false}
          rtl={false}
          pauseOnFocusLoss={false}
        />
        <FilterProvider>{children}</FilterProvider>
      </TestCommonWrapper>
    );
  }
  return await rtlRender(ui, { wrapper: AllTheProviders, ...renderOptions });
}

// override render method
export { render, renderAsync };
