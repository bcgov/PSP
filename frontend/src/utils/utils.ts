import { AxiosError } from 'axios';
import { SelectOption } from 'components/common/form';
import { TableSort } from 'components/Table/TableSort';
import { FormikProps, getIn } from 'formik';
import _ from 'lodash';
import { isEmpty, isNull, isUndefined, keys, lowerFirst, startCase } from 'lodash';
import moment, { Moment } from 'moment-timezone';
import { hideLoading, showLoading } from 'react-redux-loading-bar';
import { ILookupCode } from 'store/slices/lookupCodes';
import { logError, logRequest, logSuccess } from 'store/slices/network/networkSlice';

/**
 * Convert the specified 'input' value into a decimal or undefined.
 * @param input The string value to convert to a decimal.
 */
export const decimalOrUndefined = (input: string): number | undefined => {
  return input !== '' && input !== undefined ? parseInt(input, 10) : undefined;
};

/**
 * Convert the specified 'input' value into a float or undefined.
 * @param input The string value to convert to a float.
 */
export const floatOrUndefined = (input: string): number | undefined => {
  return input !== '' && input !== undefined ? parseFloat(input) : undefined;
};

/**
 * Determine if the specified 'input' value is a positive number of zero.
 * @param input The value to evaluate.
 * @returns True if the value is a positive number or zero, false otherwise.
 */
export const isPositiveNumberOrZero = (input: string | number | undefined | null) => {
  if (isNull(input) || isUndefined(input)) {
    return false;
  }

  if (typeof input === 'string' && isEmpty(input.trim())) {
    return false;
  }

  return !isNaN(Number(input)) && Number(input) > -1;
};

/** used for filters that need to display the string value of a parent organization organization */
export const mapLookupCodeWithParentString = (
  code: ILookupCode,
  /** the list of lookup codes to look for parent */
  options: ILookupCode[],
): SelectOption => ({
  label: code.name,
  value: code.id.toString(),
  code: code.code,
  parentId: code.parentId,
  parent: options.find((a: ILookupCode) => a.id.toString() === code.parentId?.toString())?.name,
});

/** used for inputs that need to display the string value of a parent organization organization */
export const mapSelectOptionWithParent = (
  code: SelectOption,
  /** the list of lookup codes to look for parent */
  options: SelectOption[],
): SelectOption => ({
  label: code.label,
  value: code.value.toString(),
  code: code.code,
  parentId: code.parentId,
  parent: options.find((a: SelectOption) => a.value.toString() === code.parentId?.toString())
    ?.label,
});

type FormikMemoProps = {
  formikProps: FormikProps<any>;
  field: string;
  disabled?: boolean;
  options?: SelectOption[];
} & any;
/**
 * Common use memo function prevents renders unless associated field data has been changed.
 * Essential for performance on large forms. Adapted from:
 * https://jaredpalmer.com/formik/docs/api/fastfield
 * @param param0 params from previous render
 * @param param1 params from current render
 */
export const formikFieldMemo = (
  {
    formikProps: currentProps,
    field: currField,
    disabled: currentDisabled,
    options: currentOptions,
  }: FormikMemoProps,
  {
    formikProps: prevProps,
    field: prevField,
    disabled: prevDisabled,
    options: prevOptions,
  }: FormikMemoProps,
) => {
  return !(
    currField !== prevField ||
    currentDisabled !== prevDisabled ||
    currentOptions !== prevOptions ||
    getIn(currentProps.values, prevField) !== getIn(prevProps.values, prevField) ||
    getIn(currentProps.errors, prevField) !== getIn(prevProps.errors, prevField) ||
    getIn(currentProps.touched, prevField) !== getIn(prevProps.touched, prevField) ||
    Object.keys(currentProps).length !== Object.keys(prevProps).length ||
    currentProps.isSubmitting !== prevProps.isSubmitting
  );
};

/** Provides default redux boilerplate applicable to most axios redux actions
 * @param dispatch Dispatch function
 * @param actionType All dispatched GenericNetworkActions will use this action type.
 * @param axiosPromise The result of an axios.get, .put, ..., call.
 */
export const handleAxiosResponse = (
  dispatch: Function,
  actionType: string,
  axiosPromise: Promise<any>,
): Promise<any> => {
  dispatch(logRequest(actionType));
  dispatch(showLoading());
  return axiosPromise
    .then((response: any) => {
      dispatch(logSuccess({ name: actionType }));
      dispatch(hideLoading());
      return response.data ?? response;
    })
    .catch((axiosError: AxiosError) => {
      dispatch(
        logError({ name: actionType, status: axiosError?.response?.status, error: axiosError }),
      );
      throw axiosError;
    })
    .finally(() => {
      dispatch(hideLoading());
    });
};

/**
 * convert table sort config to api sort query
 * {name: 'desc} = ['Name desc']
 */
export const generateMultiSortCriteria = (sort: TableSort<any>) => {
  if (!sort) {
    return '';
  }

  return keys(sort).map(key => `${startCase(key).replaceAll(' ', '')} ${sort[key]}`);
};

/**
 * Convert sort query string to TableSort config
 * ['Name desc'] = {name: 'desc'}
 */
export const resolveSortCriteriaFromUrl = (input: string[]): TableSort<any> | {} => {
  if (isEmpty(input)) {
    return {};
  }

  return input.reduce((acc: any, sort: string) => {
    const fields = sort.split(' ');
    if (fields.length !== 2) {
      return { ...acc };
    }

    return { ...acc, [lowerFirst(fields[0])]: fields[1] };
  }, {});
};

/**
 * get the fiscal year (ending in) based on the current date.
 */
export const getCurrentFiscalYear = (): number => {
  const now = moment();
  return now.month() >= 4 ? now.add(1, 'years').year() : now.year();
};

export const formatDate = (date?: string | Date | Moment) => {
  return !!date ? moment(date).format('YYYY-MM-DD') : '';
};

export const prettyFormatDate = (date?: string | Date | Moment) => {
  return !!date ? moment(date).format('MMM D, YYYY') : '';
};

/**
 * Format the passed string date assuming the date was recorded in UTC (as is the case with the pims API server).
 * Returns a date formatted for display in the current time zone of the user.
 * @param date utc date/time string.
 */
export const formatApiDateTime = (date?: string | Date | Moment) => {
  if (typeof date === 'string')
    return moment
      .utc(date)
      .local()
      .format('YYYY-MM-DD hh:mm a');
  return !!date
    ? moment
        .utc(date)
        .local()
        .format('YYYY-MM-DD hh:mm a')
    : '';
};

/**
 * Get the current date time in the UTC timezone. This allows the frontend to create timestamps that are compatible with timestamps created by the API.
 */
export const generateUtcNowDateTime = () =>
  moment(new Date())
    .utc()
    .format('YYYY-MM-DDTHH:mm:ss.SSSSSSS');

/**
 * Returns true only if the passed mouse event occurred within the last 500ms, or the mouse event is null.
 */
export const isMouseEventRecent = (timeStamp?: number) =>
  !!timeStamp && timeStamp >= (document?.timeline?.currentTime ?? 0) - 500;

/**
 * postalCodeFormatter takes the specified postal code and formats it with a space in the middle
 * @param {string} postal The target postal to be formatted
 */
export const postalCodeFormatter = (postal: string) => {
  const regex = /([a-zA-z][0-9][a-zA-z])[\s-]?([0-9][a-zA-z][0-9])/;
  const format = postal.match(regex);
  if (format !== null && format.length === 3) {
    postal = `${format[1]} ${format[2]}`;
  }
  return postal.toUpperCase();
};

/**
 * Using the administrative areas code set, find the matching municipality returned by the parcel layer, if present.
 * @param administrativeAreas the full list from the administrative areas code set.
 * @param layerMunicipality the municipality returned by the layer.
 */
export const getAdminAreaFromLayerData = (
  administrativeAreas: ILookupCode[],
  layerMunicipality: string,
) => {
  let administrativeArea = _.find(administrativeAreas, { name: layerMunicipality });
  if (administrativeArea) {
    return administrativeArea;
  }
  if (!!layerMunicipality?.length) {
    const splitLayerMunicipality = layerMunicipality.split(',');
    if (splitLayerMunicipality.length === 2) {
      const formattedLayerMunicipality = `${splitLayerMunicipality[1].trim()} ${splitLayerMunicipality[0].trim()}`;
      let match = _.find(administrativeAreas, { name: formattedLayerMunicipality });
      if (!match) {
        match = _.find(administrativeAreas, { name: splitLayerMunicipality[0].trim() });
      }
      return match;
    }
  }
};

/**
 * Add a simple retry wrapper to help avoid chunk errors in deployed pims application, recursively calls promise based on attemptsLeft parameter.
 * @param lazyComponent
 * @param attemptsLeft
 */
export default function componentLoader(lazyComponent: Promise<any>, attemptsLeft: number) {
  return new Promise<any>((resolve, reject) => {
    lazyComponent.then(resolve).catch((error: any) => {
      setTimeout(() => {
        if (attemptsLeft === 0) {
          reject(error);
          return;
        }
        componentLoader(lazyComponent, attemptsLeft - 1).then(resolve, reject);
      }, 500);
    });
  });
}
