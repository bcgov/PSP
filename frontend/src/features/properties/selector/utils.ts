import { IMapProperty } from 'features/properties/selector/models';
import { compact } from 'lodash';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { pidFormatter } from 'utils';

export enum NameSourceType {
  PID = 'PID',
  PIN = 'PIN',
  PLAN = 'Plan #',
  LOCATION = 'Location',
  NAME = 'Descriptive Name',
  NONE = 'n/a',
}

interface PropertyName {
  label: NameSourceType;
  value: string;
}

export const getPropertyName = (property: IMapProperty): PropertyName => {
  if (property.pid !== undefined && property.pid?.toString().length > 0 && property.pid !== '0') {
    return { label: NameSourceType.PID, value: pidFormatter(property.pid.toString()) };
  } else if (
    property.pin !== undefined &&
    property.pin?.toString()?.length > 0 &&
    property.pin !== '0'
  ) {
    return { label: NameSourceType.PIN, value: property.pin.toString() };
  } else if (property.planNumber !== undefined) {
    return { label: NameSourceType.PLAN, value: property.planNumber };
  } else if (property.latitude !== undefined && property.longitude !== undefined) {
    return {
      label: NameSourceType.LOCATION,
      value: compact([property.latitude?.toFixed(5), property.longitude?.toFixed(5)]).join(', '),
    };
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getFilePropertyName = (
  fileProperty?: Api_PropertyFile,
  skipName: boolean = false,
): PropertyName => {
  if (fileProperty === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  if (
    fileProperty.propertyName !== undefined &&
    fileProperty.propertyName !== '' &&
    skipName === false
  ) {
    return { label: NameSourceType.NAME, value: fileProperty.propertyName };
  } else if (fileProperty.property !== undefined) {
    const property = fileProperty.property;
    let mapProperty: IMapProperty = {
      pin: property.pin?.toString(),
      pid: property.pid?.toString(),
      latitude: property.latitude,
      longitude: property.longitude,
      planNumber: property.planNumber,
    };
    return getPropertyName(mapProperty);
  }
  return { label: NameSourceType.NONE, value: '' };
};
