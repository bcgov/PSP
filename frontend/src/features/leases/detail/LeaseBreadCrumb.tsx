import * as CommonStyled from 'components/common/styles';
import * as React from 'react';
import { Breadcrumb } from 'react-bootstrap';
import { Link } from 'react-router-dom';

import { ILeasePage } from './LeaseContainer';

interface ILeaseAndLicenseBreadCrumbProps {
  leaseId?: number;
  leasePage?: ILeasePage;
  onClickManagement: () => void;
}

/**
 * breadcrumb navigation for leases pages.
 * @param {ILeaseAndLicenseBreadCrumbProps} param0
 */
export const LeaseBreadCrumb: React.FunctionComponent<ILeaseAndLicenseBreadCrumbProps> = ({
  leaseId,
  leasePage,
  onClickManagement,
}) => {
  return (
    <CommonStyled.Breadcrumb>
      <Breadcrumb.Item onClick={onClickManagement}>Management</Breadcrumb.Item>
      {/* Render link only if leaseId is available */}
      {leaseId && (
        <Breadcrumb.Item linkAs={Link} linkProps={{ to: '/lease/list' }}>
          Lease &amp; License Search
        </Breadcrumb.Item>
      )}
      <Breadcrumb.Item active>{leasePage?.title}</Breadcrumb.Item>
    </CommonStyled.Breadcrumb>
  );
};

export default LeaseBreadCrumb;
