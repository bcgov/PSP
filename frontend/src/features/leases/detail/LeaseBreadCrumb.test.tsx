import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { ReactElement } from 'react';
import { render, RenderOptions } from 'utils/test-utils';

import LeaseBreadCrumb from './LeaseBreadCrumb';
import { LeasePageNames, leasePages } from './LeaseContainer';

const history = createMemoryHistory();
const onClickManagement = jest.fn();

describe('LeaseBreadCrumb component', () => {
  const setup = (renderOptions: RenderOptions & { breadcrumb?: ReactElement } = {}) => {
    // render component under test
    const component = render(
      renderOptions.breadcrumb ?? <LeaseBreadCrumb onClickManagement={onClickManagement} />,
      {
        ...renderOptions,
        history,
      },
    );

    return {
      component,
    };
  };
  beforeEach(() => {
    onClickManagement.mockReset();
  });
  it('renders as expected with no data', () => {
    const { component } = setup();
    expect(component.asFragment()).toMatchSnapshot();
  });
  it('renders as expected with data', () => {
    const { component } = setup({
      breadcrumb: (
        <LeaseBreadCrumb
          onClickManagement={onClickManagement}
          leaseId={1}
          leasePage={leasePages.get(LeasePageNames.DETAILS)}
        />
      ),
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('fires the onClickManagement event when Management section of breadcrumb is clicked.', () => {
    const { component } = setup();
    const { getByText } = component;
    userEvent.click(getByText('Management'));
    expect(onClickManagement).toHaveBeenCalled();
  });
  it('navigates to the expected location when Lease & License search clicked', () => {
    const { component } = setup();
    const { getByText } = component;
    userEvent.click(getByText('Lease & License Search'));
    expect(history.location.pathname).toBe('/lease/list');
  });

  it('the last part of the breadcrumb is active', () => {
    const { component } = setup({
      breadcrumb: (
        <LeaseBreadCrumb
          onClickManagement={onClickManagement}
          leaseId={1}
          leasePage={leasePages.get(LeasePageNames.DETAILS)}
        />
      ),
    });
    const { getByText } = component;
    const detailsLink = getByText('Details');
    expect(detailsLink).toHaveClass('active');
  });
});
