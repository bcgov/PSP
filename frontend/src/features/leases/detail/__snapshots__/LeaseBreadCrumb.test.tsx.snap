// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaseBreadCrumb component renders as expected with data 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="Toastify"
  />
  .c0 {
  grid-area: breadcrumb;
}

.c0 .breadcrumb-item:not(:first-child)::before {
  content: '>';
  width: 0px;
  margin-right: 1rem;
}

.c0 .breadcrumb li a {
  max-height: 24px;
}

.c0 ol {
  background-color: white;
}

<nav
    aria-label="breadcrumb"
    class="c0"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="breadcrumb-item"
      >
        <a
          href="#"
          role="button"
        >
          Management
        </a>
      </li>
      <li
        class="breadcrumb-item"
      >
        <a
          href="/lease/list"
        >
          Lease & License Search
        </a>
      </li>
      <li
        aria-current="page"
        class="breadcrumb-item active"
      >
        Details
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;

exports[`LeaseBreadCrumb component renders as expected with no data 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="Toastify"
  />
  .c0 {
  grid-area: breadcrumb;
}

.c0 .breadcrumb-item:not(:first-child)::before {
  content: '>';
  width: 0px;
  margin-right: 1rem;
}

.c0 .breadcrumb li a {
  max-height: 24px;
}

.c0 ol {
  background-color: white;
}

<nav
    aria-label="breadcrumb"
    class="c0"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="breadcrumb-item"
      >
        <a
          href="#"
          role="button"
        >
          Management
        </a>
      </li>
      <li
        class="breadcrumb-item"
      >
        <a
          href="/lease/list"
        >
          Lease & License Search
        </a>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;
